drivers/gpio/gpio_shell.c

```diff
@@ -107,25 +138,43 @@ static int cmd_gpio_get(const struct shell *sh,
 static int cmd_gpio_set(const struct shell *sh,
 			size_t argc, char **argv)
 {
+	const struct gpio_driver_config *cfg;
 	const struct device *dev;
 	uint8_t index = 0U;
 	uint8_t value = 0U;
+	int ret;
 
 	if (isdigit((unsigned char)argv[args_indx.index][0]) != 0 &&
 	    isdigit((unsigned char)argv[args_indx.value][0]) != 0) {
 		index = (uint8_t)atoi(argv[args_indx.index]);
 		value = (uint8_t)atoi(argv[args_indx.value]);
 	} else {
-		shell_print(sh, "Wrong parameters for set");
+		shell_print(sh, "Wrong parameters for GPIO set");
 		return -EINVAL;
 	}
 	dev = device_get_binding(argv[args_indx.port]);
 
 	if (dev != NULL) {
-		index = (uint8_t)atoi(argv[2]);
-		shell_print(sh, "Writing to %s pin %d",
+		cfg = (const struct gpio_driver_config *)dev->config;
+		if (cfg == NULL) {
+			shell_error(sh, "Device Configuration is NULL");
+			return -EIO;
+		}
+		if (index < find_msb_set(cfg->port_pin_mask)) {
+			shell_print(sh, "Writing to %s pin %u",
 			    argv[args_indx.port], index);
-		gpio_pin_set(dev, index, value);
+		} else {
+			shell_error(sh, "Invalid GPIO pin");
+			return -EINVAL;
+		}
+		ret = gpio_pin_set(dev, index, value);
+		if (ret < 0) {
+			shell_error(sh, "Error %d while writing value", ret);
+			return -EIO;
+		}
+	} else {
+		shell_error(sh, "No such GPIO device");
+		return -ENODEV;
 	}
```

Comment:
Same here - I'm curious as to why you need to reach into the implementation here and not simply use `gpio_pin_set()`

Reply:
The changes are needed if a zephyr application has enabled gpio shell. If a user passes a invalid pin number which is more then actual number of pins available,  then following results is observed currently :
agilex$ gpio get gpio@10c03200 24
Reading gpio@10c03200 pin 24
Value 0 
while valid pin values are from 0 to 23 only.

Fix is added in gpio_shell.c for 3 reasons :
1. Few API we don't pass pin values directly to the driver api. For example, gpio_pin_get will get whole port value and then do mask :
  int (*port_get_raw)(const struct device *port,
			      gpio_port_value_t *value);

	  int (*port_set_masked_raw)(const struct device *port,
				     gpio_port_pins_t mask,
				     gpio_port_value_t value);
2. All the driver might need to have same functionality when shell is implemented.
3. Values passed above 32, will always give value 0 even if currently no. of pins above 32 doesn't exits.

New driver doesn't this this change for valid pins. This is just for invalid pins.